elastic-apm-agent/src/main/java/co/elastic/apm/agent/premain/ShadedClassLoader.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package co.elastic.apm.agent.premain;
+
+import javax.annotation.Nullable;
+import java.io.BufferedInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.util.Enumeration;
+import java.util.jar.JarFile;
+import java.util.jar.Manifest;
+
+public class ShadedClassLoader extends URLClassLoader {
+
+    private static final String CLASS_EXTENSION = ".class";
+    private final String prefix;
+    private final String classFileExtension;
+    private final Manifest manifest;
+    private final URL agentJarUrl;
+
+    static {
+        ClassLoader.registerAsParallelCapable();
+    }
+
+    public ShadedClassLoader(File agentJar, ClassLoader parent, String prefix, String classNameExtension) throws IOException {
+        super(new URL[]{agentJar.toURI().toURL()}, parent);
+        this.prefix = prefix;
+        this.classFileExtension = classNameExtension;
+        this.agentJarUrl = agentJar.toURI().toURL();
+        try (JarFile jarFile = new JarFile(agentJar)) {
+            this.manifest = jarFile.getManifest();
+        }
+    }
+
+    @Override
+    protected Class<?> findClass(String name) throws ClassNotFoundException {
+        try {
+            return super.findClass(name);
+        } catch (ClassNotFoundException e) {
+            byte[] classBytes = getClassBytes(name);
+            if (classBytes != null) {
+                return defineClass(name, classBytes);
+            } else {
+                throw new ClassNotFoundException(name);
+            }
+        }
+    }
+
+    private Class<?> defineClass(String name, byte[] classBytes) {
+        String packageName = getPackageName(name);
+        if (packageName != null && getPackage(packageName) == null) {
+            synchronized (this) {
```

Comment:
That's because multiple classes may cause a single package to be defined. The first class of a given package that's loaded will define the package.

Reply:
Sorry, I wasn't very clear - I asked whether it is the right lock to use, rather than the one returned from `getClassLoadingLock`. I had more time to look into the implementation now and it seems it is the right one - the per-class-name lock is synchronized through `loadClass`, so in for this scope it indeed seems the proper lock.